### PR TITLE
Add support for retrieving mapped externalClaims for given localClaim

### PR DESCRIPTION
**Pull request by ChanikaRuchini**
### Proposed changes in this pull request

This PR provides support at claimmetadatamgtservice to fetch the mapped external claims of a local claim

Related issue
- https://github.com/wso2/product-is/issues/16702

### CODE REVIEW

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedExternalClaimDAO.java [hunk 1]
```diff
@@ -72,20 +75,32 @@ public void addExternalClaim(ExternalClaim externalClaim, int tenantId) throws C
...
         externalClaimDAO.updateExternalClaim(externalClaim, tenantId);
         String externalClaimDialectURI = externalClaim.getClaimDialectURI();
         ExternalClaimCacheKey cacheKey = new ExternalClaimCacheKey(externalClaimDialectURI);
         externalClaimCache.clearCacheEntry(cacheKey, tenantId);
+        associatedClaimCache.clearCacheEntry(externalClaim.getMappedLocalClaim(), tenantId);
     }
     public void removeExternalClaim(String externalClaimDialectURI, String externalClaimURI, int tenantId) throws
             ClaimMetadataException {
 
+        List<ExternalClaim> externalClaimsList = getExternalClaims(externalClaimDialectURI, tenantId);
```
Comment (AnuradhaSK) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedExternalClaimDAO.java:91]: If `externalClaimsList` is nullable, check not null before looping

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java [hunk 2]
```diff
@@ -368,4 +368,34 @@ private static void rollbackTransaction(Connection dbConnection) {
...
         }
     }
+
+    /**
+     * Fetch mapped external claims of a local claim.
+     *
+     * @param tenantId      Tenant Id.
+     * @param localClaimURI URI of the local claim.
+     * @return List of associated external claims.
+     * @throws ClaimMetadataException when trying to fetch mapped external claims for a local claim.
```
Comment (AnuradhaSK) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java:378]:
```suggestion
     * @throws ClaimMetadataException When trying to fetch mapped external claims for a local claim.
```

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java [hunk 3]
```diff
@@ -368,4 +368,34 @@ private static void rollbackTransaction(Connection dbConnection) {
...
+     */
+    public List<Claim> fetchMappedExternalClaims(String localClaimURI, int tenantId)
+            throws ClaimMetadataException {
+
+        List<Claim> mappedExternalClaims = new ArrayList<>();
+        try (Connection dbConnection = IdentityDatabaseUtil.getDBConnection(true);
```
Comment (AnuradhaSK) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java:384]: Transaction is not needed here

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java [hunk 4]
```diff
@@ -115,4 +121,31 @@ public void removeClaimMappingAttributes(int tenantId, String userstoreDomain) t
...
+     * @param localClaimURI URI of local claim.
+     * @throws ClaimMetadataException If an error occurred while getting mapped external claim for local claim.
+     */
+    public List<Claim> fetchMappedExternalClaims(String localClaimURI, int tenantId)
+            throws ClaimMetadataException {
+
+        List<Claim> associatedLocalClaims = associatedClaimCache.getValueFromCache(localClaimURI, tenantId);
+        if (associatedLocalClaims == null) {
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Cache miss for associated claims of local claim:%s for tenant:%s ",
```
Comment (AnuradhaSK) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java:138]:
```suggestion
                log.debug(String.format("Cache miss for associated claims of local claim:%s in tenant:%s ",
```

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java [hunk 5]
```diff
@@ -115,4 +121,31 @@ public void removeClaimMappingAttributes(int tenantId, String userstoreDomain) t
...
+                        localClaimURI, tenantId));
+            }
+            associatedLocalClaims = localClaimDAO.fetchMappedExternalClaims(localClaimURI, tenantId);
+            associatedClaimCache.addToCache(localClaimURI, new ArrayList<>(associatedLocalClaims), tenantId);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Cache hit for associated claims of local claim:%s for tenant:%s ",
```
Comment (AnuradhaSK) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java:145]:
```suggestion
                log.debug(String.format("Cache hit for associated claims of local claim:%s in tenant:%s ",
```

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/cache/AssociatedClaimCache.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com) All Rights Reserved.
```
Comment (sadilchamishka) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/cache/AssociatedClaimCache.java:2]: Check the license

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedExternalClaimDAO.java (hunk 1 above)
Comment (ChanikaRuchini) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedExternalClaimDAO.java:91]: Fixed

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java (hunk 2 above)
Comment (ChanikaRuchini) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java]: Fixed

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java (hunk 3 above)
Comment (ChanikaRuchini) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java]: Fixed

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java (hunk 4 above)
Comment (ChanikaRuchini) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java]: Fixed

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java (hunk 5 above)
Comment (ChanikaRuchini) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java]: Fixed

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/cache/AssociatedClaimCache.java (hunk 6 above)
Comment (ChanikaRuchini) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/cache/AssociatedClaimCache.java]: Fixed

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java [hunk 7]
```diff
@@ -207,4 +208,18 @@ default String getMaskingRegexForLocalClaim(String localClaimURI, String tenantD
...
         return null;
     }
+
+    /**
+     * Get mapped external claims for provided local claim.
+     *
+     * @param localClaimURI Local claim URI.
+     * @param tenantDomain  Tenant domain.
+     * @return List of mapped external claims.
+     * @throws ClaimMetadataException If an error occurred while retrieving local claims.
```
Comment (AnuradhaSK) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java:218]:
```suggestion
     * @throws ClaimMetadataException If an error occurred while retrieving external claims.
```

Comment (ChanikaRuchini) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java]: Fixed

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**